File: tests/e2e/saleads_mi_negocio_full_test.py
#!/usr/bin/env python3
"""End-to-end workflow validation for SaleADS Mi Negocio module."""

from __future__ import annotations

import json
import os
import re
import sys
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable

from playwright.sync_api import (
    BrowserContext,
    Locator,
    Page,
    TimeoutError as PlaywrightTimeoutError,
    sync_playwright,
)


DEFAULT_TIMEOUT_MS = int(os.getenv("SALEADS_TIMEOUT_MS", "20000"))
HEADLESS = os.getenv("HEADLESS", "false").strip().lower() in {"1", "true", "yes"}
BASE_URL = os.getenv("SALEADS_BASE_URL", "").strip()
GOOGLE_ACCOUNT_EMAIL = os.getenv(
    "GOOGLE_ACCOUNT_EMAIL", "juanlucasbarbiergarzon@gmail.com"
).strip()
RUN_ID = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
ARTIFACTS_DIR = Path(
    os.getenv("SALEADS_ARTIFACTS_DIR", f"tests/e2e/artifacts/{RUN_ID}")
).resolve()
SCREENSHOTS_DIR = ARTIFACTS_DIR / "screenshots"


@dataclass
class ValidationResult:
    description: str
    passed: bool
    details: str = ""


@dataclass
class StepResult:
    name: str
    validations: list[ValidationResult] = field(default_factory=list)
    evidence: dict[str, object] = field(default_factory=dict)
    error: str = ""
    passed: bool = False

    def add_validation(self, description: str, passed: bool, details: str = "") -> None:
        self.validations.append(
            ValidationResult(description=description, passed=passed, details=details)
        )

    def finalize(self) -> None:
        if self.error:
            self.passed = False
            return

        if not self.validations:
            self.passed = True
            return

        self.passed = all(result.passed for result in self.validations)


def wait_for_ui(page: Page) -> None:
    for state in ("domcontentloaded", "networkidle"):
        try:
            page.wait_for_load_state(state, timeout=DEFAULT_TIMEOUT_MS)
        except PlaywrightTimeoutError:
            # Some screens keep persistent network activity. Do not block the flow.
            continue


def text_patterns(texts: Iterable[str]) -> list[re.Pattern[str]]:
    return [re.compile(re.escape(text), re.IGNORECASE) for text in texts]


def locator_is_visible(locator: Locator, timeout_ms: int = 3500) -> bool:
    try:
        locator.first.wait_for(state="visible", timeout=timeout_ms)
        return True
    except PlaywrightTimeoutError:
        return False


def find_target(page: Page, texts: Iterable[str]) -> Locator:
    patterns = text_patterns(texts)
    role_candidates = ("button", "link", "menuitem", "tab")

    for pattern in patterns:
        for role in role_candidates:
            role_locator = page.get_by_role(role, name=pattern)
            if locator_is_visible(role_locator):
                return role_locator.first

        text_locator = page.get_by_text(pattern)
        if locator_is_visible(text_locator):
            return text_locator.first

    joined = ", ".join(texts)
    raise RuntimeError(f"Could not find visible element with text: {joined}")


def safe_click(page: Page, locator: Locator) -> None:
    locator.click(timeout=DEFAULT_TIMEOUT_MS)
    wait_for_ui(page)


def click_by_text(page: Page, texts: Iterable[str]) -> Locator:
    target = find_target(page, texts)
    safe_click(page, target)
    return target


def is_any_text_visible(page: Page, texts: Iterable[str], timeout_ms: int = 4500) -> bool:
    for pattern in text_patterns(texts):
        if locator_is_visible(page.get_by_text(pattern), timeout_ms=timeout_ms):
            return True
    return False


def is_heading_visible(page: Page, texts: Iterable[str], timeout_ms: int = 4500) -> bool:
    for pattern in text_patterns(texts):
        heading = page.get_by_role("heading", name=pattern)
        if locator_is_visible(heading, timeout_ms=timeout_ms):
            return True
        if locator_is_visible(page.get_by_text(pattern), timeout_ms=timeout_ms):
            return True
    return False


def find_business_name_input(page: Page) -> Locator | None:
    candidates = [
        page.get_by_label(re.compile(r"Nombre del Negocio", re.IGNORECASE)),
        page.get_by_placeholder(re.compile(r"Nombre del Negocio", re.IGNORECASE)),
        page.locator("input[name*='negocio' i], input[id*='negocio' i]"),
    ]
    for candidate in candidates:
        if locator_is_visible(candidate, timeout_ms=2500):
            return candidate.first
    return None


def capture_screenshot(
    page: Page, step: StepResult, file_name: str, full_page: bool = False
) -> None:
    SCREENSHOTS_DIR.mkdir(parents=True, exist_ok=True)
    screenshot_path = SCREENSHOTS_DIR / file_name
    page.screenshot(path=str(screenshot_path), full_page=full_page)
    current = step.evidence.setdefault("screenshots", [])
    if isinstance(current, list):
        current.append(str(screenshot_path))


def get_sidebar_visible(page: Page) -> bool:
    aside_or_nav = page.locator("aside, nav")
    try:
        aside_or_nav.first.wait_for(state="visible", timeout=DEFAULT_TIMEOUT_MS)
        return True
    except PlaywrightTimeoutError:
        return False


def bootstrap(context: BrowserContext) -> Page:
    page = context.new_page()
    page.set_default_timeout(DEFAULT_TIMEOUT_MS)

    if BASE_URL:
        page.goto(BASE_URL, wait_until="domcontentloaded")
        wait_for_ui(page)
    else:
        # Prompt allows any environment; URL comes from variable to avoid hardcoding domains.
        page.goto("about:blank")
    return page


def step_login_with_google(page: Page) -> StepResult:
    step = StepResult(name="Login")
    try:
        if not BASE_URL:
            raise RuntimeError(
                "SALEADS_BASE_URL is not set. Provide the environment login URL."
            )

        login_texts = (
            "Sign in with Google",
            "Iniciar con Google",
            "Iniciar sesión con Google",
            "Continuar con Google",
            "Google",
        )
        login_target = find_target(page, login_texts)

        popup_page: Page | None = None
        try:
            with page.expect_popup(timeout=5000) as popup_info:
                safe_click(page, login_target)
            popup_page = popup_info.value
            wait_for_ui(popup_page)
        except PlaywrightTimeoutError:
            safe_click(page, login_target)

        google_page = popup_page if popup_page else page
        if is_any_text_visible(google_page, (GOOGLE_ACCOUNT_EMAIL,), timeout_ms=5000):
            click_by_text(google_page, (GOOGLE_ACCOUNT_EMAIL,))

        if popup_page is not None:
            try:
                popup_page.wait_for_event("close", timeout=20000)
            except PlaywrightTimeoutError:
                # Continue even if popup remains open due to account already selected.
                pass
            page.bring_to_front()
            wait_for_ui(page)

        main_interface_visible = is_any_text_visible(
            page,
            ("Dashboard", "Panel", "Inicio", "Mi Negocio", "Negocio"),
            timeout_ms=15000,
        )
        sidebar_visible = get_sidebar_visible(page)

        step.add_validation(
            "Confirm the main application interface appears.",
            main_interface_visible,
            "Expected dashboard or navigation text after login.",
        )
        step.add_validation(
            "Confirm the left sidebar navigation is visible.",
            sidebar_visible,
            "Expected an aside/nav container visible in the app shell.",
        )
        capture_screenshot(page, step, "01_dashboard_loaded.png")
    except Exception as error:  # pragma: no cover - defensive flow reporting
        step.error = str(error)

    step.finalize()
    return step


def step_open_mi_negocio_menu(page: Page) -> StepResult:
    step = StepResult(name="Mi Negocio menu")
    try:
        if is_any_text_visible(page, ("Negocio",), timeout_ms=5000):
            click_by_text(page, ("Negocio",))

        click_by_text(page, ("Mi Negocio",))

        submenu_expanded = is_any_text_visible(
            page, ("Agregar Negocio", "Administrar Negocios"), timeout_ms=10000
        )
        add_business_visible = is_any_text_visible(page, ("Agregar Negocio",))
        manage_business_visible = is_any_text_visible(page, ("Administrar Negocios",))

        step.add_validation("Confirm the submenu expands.", submenu_expanded)
        step.add_validation(
            "Confirm 'Agregar Negocio' is visible.", add_business_visible
        )
        step.add_validation(
            "Confirm 'Administrar Negocios' is visible.", manage_business_visible
        )
        capture_screenshot(page, step, "02_mi_negocio_expanded_menu.png")
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def step_validate_agregar_negocio_modal(page: Page) -> StepResult:
    step = StepResult(name="Agregar Negocio modal")
    try:
        click_by_text(page, ("Agregar Negocio",))

        modal_visible = is_heading_visible(page, ("Crear Nuevo Negocio",), timeout_ms=10000)
        business_name_input = find_business_name_input(page)
        input_visible = business_name_input is not None
        quota_visible = is_any_text_visible(page, ("Tienes 2 de 3 negocios",))
        cancel_visible = is_any_text_visible(page, ("Cancelar",))
        create_visible = is_any_text_visible(page, ("Crear Negocio",))

        step.add_validation("Modal title 'Crear Nuevo Negocio' is visible.", modal_visible)
        step.add_validation(
            "Input field 'Nombre del Negocio' exists.", input_visible
        )
        step.add_validation(
            "Text 'Tienes 2 de 3 negocios' is visible.", quota_visible
        )
        step.add_validation("Buttons 'Cancelar' and 'Crear Negocio' are present.", cancel_visible and create_visible)

        capture_screenshot(page, step, "03_agregar_negocio_modal.png")

        if business_name_input is not None:
            try:
                business_name_input.fill("Negocio Prueba Automatización")
            except Exception:
                # Optional action only.
                pass

        if cancel_visible:
            click_by_text(page, ("Cancelar",))
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def step_open_administrar_negocios(page: Page) -> StepResult:
    step = StepResult(name="Administrar Negocios view")
    try:
        if not is_any_text_visible(page, ("Administrar Negocios",), timeout_ms=2000):
            click_by_text(page, ("Mi Negocio",))

        click_by_text(page, ("Administrar Negocios",))

        info_general_visible = is_any_text_visible(page, ("Información General",), timeout_ms=15000)
        account_details_visible = is_any_text_visible(page, ("Detalles de la Cuenta",), timeout_ms=10000)
        business_section_visible = is_any_text_visible(page, ("Tus Negocios",), timeout_ms=10000)
        legal_section_visible = is_any_text_visible(page, ("Sección Legal",), timeout_ms=10000)

        step.add_validation("Section 'Información General' exists.", info_general_visible)
        step.add_validation("Section 'Detalles de la Cuenta' exists.", account_details_visible)
        step.add_validation("Section 'Tus Negocios' exists.", business_section_visible)
        step.add_validation("Section 'Sección Legal' exists.", legal_section_visible)
        capture_screenshot(page, step, "04_administrar_negocios_page.png", full_page=True)
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def step_validate_informacion_general(page: Page) -> StepResult:
    step = StepResult(name="Información General")
    try:
        user_name_visible = is_any_text_visible(
            page,
            ("@", "Nombre", "Usuario", "Perfil"),
            timeout_ms=6000,
        )
        user_email_visible = is_any_text_visible(page, ("@", ".com", ".ai"), timeout_ms=6000)
        business_plan_visible = is_any_text_visible(page, ("BUSINESS PLAN",))
        change_plan_visible = is_any_text_visible(page, ("Cambiar Plan",))

        step.add_validation("User name is visible.", user_name_visible)
        step.add_validation("User email is visible.", user_email_visible)
        step.add_validation("Text 'BUSINESS PLAN' is visible.", business_plan_visible)
        step.add_validation("Button 'Cambiar Plan' is visible.", change_plan_visible)
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def step_validate_detalles_cuenta(page: Page) -> StepResult:
    step = StepResult(name="Detalles de la Cuenta")
    try:
        created_visible = is_any_text_visible(page, ("Cuenta creada",))
        active_visible = is_any_text_visible(page, ("Estado activo",))
        language_visible = is_any_text_visible(page, ("Idioma seleccionado",))

        step.add_validation("'Cuenta creada' is visible.", created_visible)
        step.add_validation("'Estado activo' is visible.", active_visible)
        step.add_validation("'Idioma seleccionado' is visible.", language_visible)
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def step_validate_tus_negocios(page: Page) -> StepResult:
    step = StepResult(name="Tus Negocios")
    try:
        list_visible = is_any_text_visible(page, ("Tus Negocios", "Negocio"), timeout_ms=8000)
        add_button_visible = is_any_text_visible(page, ("Agregar Negocio",))
        quota_visible = is_any_text_visible(page, ("Tienes 2 de 3 negocios",))

        step.add_validation("Business list is visible.", list_visible)
        step.add_validation("Button 'Agregar Negocio' exists.", add_button_visible)
        step.add_validation(
            "Text 'Tienes 2 de 3 negocios' is visible.", quota_visible
        )
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def validate_legal_content(target_page: Page, expected_heading: str) -> tuple[bool, bool]:
    heading_visible = is_heading_visible(target_page, (expected_heading,), timeout_ms=15000)
    legal_content_visible = False

    if locator_is_visible(target_page.locator("main, article, section, p"), timeout_ms=6000):
        legal_content_visible = True
    else:
        try:
            body_text = target_page.locator("body").inner_text(timeout=6000)
            legal_content_visible = len(body_text.strip()) > 150
        except Exception:
            legal_content_visible = False

    return heading_visible, legal_content_visible


def navigate_legal_link(
    page: Page,
    step_name: str,
    link_text: str,
    expected_heading: str,
    screenshot_name: str,
) -> StepResult:
    step = StepResult(name=step_name)
    app_page = page

    try:
        popup_page: Page | None = None
        target_page: Page = app_page
        current_url = app_page.url

        link_target = find_target(app_page, (link_text,))
        try:
            with app_page.expect_popup(timeout=5000) as popup_info:
                safe_click(app_page, link_target)
            popup_page = popup_info.value
            target_page = popup_page
            wait_for_ui(target_page)
        except PlaywrightTimeoutError:
            safe_click(app_page, link_target)
            wait_for_ui(app_page)
            target_page = app_page

        heading_visible, content_visible = validate_legal_content(
            target_page, expected_heading
        )
        step.add_validation(
            f"The page contains the heading '{expected_heading}'.", heading_visible
        )
        step.add_validation("Legal content text is visible.", content_visible)

        capture_screenshot(target_page, step, screenshot_name, full_page=True)
        step.evidence["final_url"] = target_page.url
        step.evidence["opened_new_tab"] = popup_page is not None

        if popup_page is not None:
            popup_page.close()
            app_page.bring_to_front()
            wait_for_ui(app_page)
        elif app_page.url != current_url:
            app_page.go_back(wait_until="domcontentloaded")
            wait_for_ui(app_page)
    except Exception as error:
        step.error = str(error)

    step.finalize()
    return step


def write_report(results: list[StepResult]) -> Path:
    ARTIFACTS_DIR.mkdir(parents=True, exist_ok=True)
    report_path = ARTIFACTS_DIR / "final_report.json"

    report = {
        "test_name": "saleads_mi_negocio_full_test",
        "run_id": RUN_ID,
        "timestamp_utc": datetime.now(timezone.utc).isoformat(),
        "base_url": BASE_URL,
        "results": {step.name: asdict(step) for step in results},
        "summary": {
            step.name: "PASS" if step.passed else "FAIL" for step in results
        },
        "all_passed": all(step.passed for step in results),
    }

    report_path.write_text(json.dumps(report, indent=2, ensure_ascii=False), encoding="utf-8")
    return report_path


def skipped_step(
    name: str, validations: list[str], reason: str
) -> StepResult:
    step = StepResult(name=name, error=reason)
    for description in validations:
        step.add_validation(description, False, reason)
    step.finalize()
    return step


def print_summary(results: list[StepResult], report_path: Path) -> None:
    print("\nFinal Report")
    for step in results:
        status = "PASS" if step.passed else "FAIL"
        print(f"- {step.name}: {status}")
        if step.error:
            print(f"  error: {step.error}")

    print(f"\nArtifacts directory: {ARTIFACTS_DIR}")
    print(f"Report JSON: {report_path}")


def main() -> int:
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=HEADLESS)
        context = browser.new_context()

        try:
            page = bootstrap(context)
            results: list[StepResult] = []

            login_step = step_login_with_google(page)
            results.append(login_step)

            if login_step.passed:
                results.extend(
                    [
                        step_open_mi_negocio_menu(page),
                        step_validate_agregar_negocio_modal(page),
                        step_open_administrar_negocios(page),
                        step_validate_informacion_general(page),
                        step_validate_detalles_cuenta(page),
                        step_validate_tus_negocios(page),
                        navigate_legal_link(
                            page=page,
                            step_name="Términos y Condiciones",
                            link_text="Términos y Condiciones",
                            expected_heading="Términos y Condiciones",
                            screenshot_name="08_terminos_y_condiciones.png",
                        ),
                        navigate_legal_link(
                            page=page,
                            step_name="Política de Privacidad",
                            link_text="Política de Privacidad",
                            expected_heading="Política de Privacidad",
                            screenshot_name="09_politica_de_privacidad.png",
                        ),
                    ]
                )
            else:
                skip_reason = (
                    "Blocked by failed login step; downstream workflow not executable."
                )
                results.extend(
                    [
                        skipped_step(
                            "Mi Negocio menu",
                            [
                                "Confirm the submenu expands.",
                                "Confirm 'Agregar Negocio' is visible.",
                                "Confirm 'Administrar Negocios' is visible.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Agregar Negocio modal",
                            [
                                "Modal title 'Crear Nuevo Negocio' is visible.",
                                "Input field 'Nombre del Negocio' exists.",
                                "Text 'Tienes 2 de 3 negocios' is visible.",
                                "Buttons 'Cancelar' and 'Crear Negocio' are present.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Administrar Negocios view",
                            [
                                "Section 'Información General' exists.",
                                "Section 'Detalles de la Cuenta' exists.",
                                "Section 'Tus Negocios' exists.",
                                "Section 'Sección Legal' exists.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Información General",
                            [
                                "User name is visible.",
                                "User email is visible.",
                                "Text 'BUSINESS PLAN' is visible.",
                                "Button 'Cambiar Plan' is visible.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Detalles de la Cuenta",
                            [
                                "'Cuenta creada' is visible.",
                                "'Estado activo' is visible.",
                                "'Idioma seleccionado' is visible.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Tus Negocios",
                            [
                                "Business list is visible.",
                                "Button 'Agregar Negocio' exists.",
                                "Text 'Tienes 2 de 3 negocios' is visible.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Términos y Condiciones",
                            [
                                "The page contains the heading 'Términos y Condiciones'.",
                                "Legal content text is visible.",
                            ],
                            skip_reason,
                        ),
                        skipped_step(
                            "Política de Privacidad",
                            [
                                "The page contains the heading 'Política de Privacidad'.",
                                "Legal content text is visible.",
                            ],
                            skip_reason,
                        ),
                    ]
                )
            report_path = write_report(results)
            print_summary(results, report_path)
            return 0 if all(step.passed for step in results) else 1
        finally:
            context.close()
            browser.close()


if __name__ == "__main__":
    sys.exit(main())

File: e2e/saleads-mi-negocio/scripts/saleads-mi-negocio-full-test.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "playwright";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_DIR = path.resolve(__dirname, "..");
const RUN_STAMP = new Date().toISOString().replace(/[:.]/g, "-");
const ARTIFACTS_DIR = path.join(PROJECT_DIR, "artifacts", RUN_STAMP);
const SCREENSHOTS_DIR = path.join(ARTIFACTS_DIR, "screenshots");

const REPORT_FIELDS = [
  "Login",
  "Mi Negocio menu",
  "Agregar Negocio modal",
  "Administrar Negocios view",
  "Información General",
  "Detalles de la Cuenta",
  "Tus Negocios",
  "Términos y Condiciones",
  "Política de Privacidad",
];

const results = Object.fromEntries(REPORT_FIELDS.map((field) => [field, "NOT_RUN"]));
const stepErrors = {};
const screenshotPaths = {};
const legalUrls = {};

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

async function waitForUi(page, timeout = 30000) {
  await page.waitForLoadState("domcontentloaded", { timeout });
  await page.waitForLoadState("networkidle", { timeout }).catch(() => {});
  await page.waitForTimeout(500);
}

function sanitizeForFileName(value) {
  return value
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-zA-Z0-9_-]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .toLowerCase();
}

async function captureCheckpoint(page, checkpointName, options = {}) {
  await ensureDir(SCREENSHOTS_DIR);
  const fileName = `${sanitizeForFileName(checkpointName)}.png`;
  const screenshotPath = path.join(SCREENSHOTS_DIR, fileName);
  await page.screenshot({ path: screenshotPath, fullPage: Boolean(options.fullPage) });
  screenshotPaths[checkpointName] = screenshotPath;
}

async function isVisible(locator) {
  return locator.isVisible().catch(() => false);
}

async function clickByVisibleText(page, matcher, intentLabel) {
  const roleLocator = page.getByRole("button", { name: matcher }).first();
  if (await isVisible(roleLocator)) {
    await roleLocator.click();
    await waitForUi(page);
    return;
  }

  const linkLocator = page.getByRole("link", { name: matcher }).first();
  if (await isVisible(linkLocator)) {
    await linkLocator.click();
    await waitForUi(page);
    return;
  }

  const textLocator = page.getByText(matcher).first();
  if (await isVisible(textLocator)) {
    await textLocator.click();
    await waitForUi(page);
    return;
  }

  throw new Error(`Could not find clickable element for: ${intentLabel}`);
}

async function waitForText(page, matcher, label, timeout = 30000) {
  const locator = page.getByText(matcher).first();
  await locator.waitFor({ state: "visible", timeout });
  return locator;
}

async function waitForAny(page, candidates, label, timeout = 30000) {
  const start = Date.now();
  while (Date.now() - start < timeout) {
    for (const factory of candidates) {
      const locator = factory();
      if (await isVisible(locator.first())) {
        return locator.first();
      }
    }
    await page.waitForTimeout(200);
  }
  throw new Error(`Could not find visible element: ${label}`);
}

async function runStep(stepName, stepAction) {
  try {
    await stepAction();
    results[stepName] = "PASS";
    console.log(`STEP PASS: ${stepName}`);
  } catch (error) {
    results[stepName] = "FAIL";
    stepErrors[stepName] = error instanceof Error ? error.message : String(error);
    console.error(`STEP FAIL: ${stepName} -> ${stepErrors[stepName]}`);
  }
}

async function maybeSelectGoogleAccount(page, email) {
  const accountLocator = page.getByText(email).first();
  try {
    await accountLocator.waitFor({ state: "visible", timeout: 10000 });
    await accountLocator.click();
    await waitForUi(page, 60000);
  } catch {
    console.log("Google account selector not shown; continuing with current session flow.");
  }
}

async function validateLegalPage({
  appPage,
  linkText,
  headingRegex,
  reportField,
  screenshotLabel,
}) {
  const popupPromise = appPage.waitForEvent("popup", { timeout: 8000 }).catch(() => null);
  await clickByVisibleText(appPage, new RegExp(linkText, "i"), linkText);
  const popupPage = await popupPromise;
  const targetPage = popupPage ?? appPage;

  await targetPage.waitForLoadState("domcontentloaded", { timeout: 30000 });
  await targetPage.waitForLoadState("networkidle", { timeout: 30000 }).catch(() => {});

  const heading = await waitForAny(
    targetPage,
    [
      () => targetPage.getByRole("heading", { name: headingRegex }),
      () => targetPage.getByText(headingRegex),
    ],
    `${linkText} heading`,
    30000,
  );
  await heading.waitFor({ state: "visible", timeout: 30000 });

  const legalText = (await targetPage.locator("main, article, body").first().innerText().catch(() => "")) || "";
  if (legalText.trim().length < 80) {
    throw new Error(`Legal content for "${linkText}" appears empty.`);
  }

  legalUrls[reportField] = targetPage.url();
  await captureCheckpoint(targetPage, screenshotLabel, { fullPage: true });

  if (popupPage) {
    await popupPage.close();
    await appPage.bringToFront();
    await waitForUi(appPage);
    return;
  }

  await appPage.goBack({ waitUntil: "domcontentloaded" }).catch(() => {});
  await waitForUi(appPage);
}

async function main() {
  await ensureDir(ARTIFACTS_DIR);
  await ensureDir(SCREENSHOTS_DIR);

  const loginUrl = process.env.SALEADS_LOGIN_URL || process.env.SALEADS_URL || "";
  const headless = process.env.HEADLESS !== "false";
  const userDataDir =
    process.env.PLAYWRIGHT_USER_DATA_DIR || path.join(PROJECT_DIR, ".pw-user-data");

  const context = await chromium.launchPersistentContext(userDataDir, {
    headless,
    viewport: { width: 1440, height: 900 },
  });

  const page = context.pages()[0] || (await context.newPage());

  try {
    if (loginUrl) {
      await page.goto(loginUrl, { waitUntil: "domcontentloaded", timeout: 60000 });
      await waitForUi(page);
    } else if (page.url().startsWith("about:blank")) {
      throw new Error(
        "No URL available. Set SALEADS_LOGIN_URL (or SALEADS_URL), or launch with a preloaded SaleADS login page.",
      );
    }

    await runStep("Login", async () => {
      await clickByVisibleText(page, /google/i, "Login with Google");
      await maybeSelectGoogleAccount(page, "juanlucasbarbiergarzon@gmail.com");

      await waitForAny(
        page,
        [() => page.locator("aside"), () => page.locator("nav"), () => page.getByText(/Negocio/i)],
        "main application interface",
        60000,
      );
      await waitForAny(
        page,
        [() => page.locator("aside"), () => page.locator("nav")],
        "left sidebar navigation",
        60000,
      );
      await captureCheckpoint(page, "01-dashboard-loaded", { fullPage: true });
    });

    await runStep("Mi Negocio menu", async () => {
      await waitForAny(
        page,
        [() => page.locator("aside"), () => page.locator("nav")],
        "left sidebar",
        30000,
      );
      await clickByVisibleText(page, /Negocio/i, "Negocio section");
      await clickByVisibleText(page, /Mi Negocio/i, "Mi Negocio option");

      await waitForText(page, /Agregar Negocio/i, "Agregar Negocio");
      await waitForText(page, /Administrar Negocios/i, "Administrar Negocios");
      await captureCheckpoint(page, "02-mi-negocio-menu-expanded");
    });

    await runStep("Agregar Negocio modal", async () => {
      await clickByVisibleText(page, /Agregar Negocio/i, "Agregar Negocio");
      await waitForText(page, /Crear Nuevo Negocio/i, "Crear Nuevo Negocio modal title");
      const dialog = await waitForAny(
        page,
        [() => page.getByRole("dialog"), () => page.locator("[role='dialog']"), () => page.locator("form")],
        "Agregar Negocio dialog",
      );

      const nameInput = await waitForAny(
        page,
        [
          () => page.getByLabel(/Nombre del Negocio/i),
          () => page.getByPlaceholder(/Nombre del Negocio/i),
          () => dialog.locator("input"),
        ],
        "Nombre del Negocio input",
      );

      await waitForText(page, /Tienes\s+2\s+de\s+3\s+negocios/i, "Business quota text");
      await waitForAny(
        page,
        [() => page.getByRole("button", { name: /Cancelar/i }), () => page.getByText(/Cancelar/i)],
        "Cancelar button",
      );
      await waitForAny(
        page,
        [() => page.getByRole("button", { name: /Crear Negocio/i }), () => page.getByText(/Crear Negocio/i)],
        "Crear Negocio button",
      );

      await captureCheckpoint(page, "03-agregar-negocio-modal");
      await nameInput.click();
      await nameInput.fill("Negocio Prueba Automatización");
      await clickByVisibleText(page, /Cancelar/i, "Cancelar");
    });

    await runStep("Administrar Negocios view", async () => {
      const administrarVisible = await isVisible(page.getByText(/Administrar Negocios/i).first());
      if (!administrarVisible) {
        await clickByVisibleText(page, /Mi Negocio/i, "Mi Negocio option");
      }

      await clickByVisibleText(page, /Administrar Negocios/i, "Administrar Negocios");
      await waitForText(page, /Información General/i, "Información General");
      await waitForText(page, /Detalles de la Cuenta/i, "Detalles de la Cuenta");
      await waitForText(page, /Tus Negocios/i, "Tus Negocios");
      await waitForText(page, /Sección Legal/i, "Sección Legal");
      await captureCheckpoint(page, "04-administrar-negocios-view", { fullPage: true });
    });

    await runStep("Información General", async () => {
      await waitForAny(
        page,
        [
          () => page.locator("section").filter({ hasText: /Información General/i }),
          () => page.getByText(/Información General/i),
        ],
        "Información General section",
      );
      await waitForAny(
        page,
        [() => page.getByText(/@/), () => page.getByText(/BUSINESS PLAN/i)],
        "user profile details",
      );
      await waitForText(page, /BUSINESS PLAN/i, "BUSINESS PLAN text");
      await waitForAny(
        page,
        [() => page.getByRole("button", { name: /Cambiar Plan/i }), () => page.getByText(/Cambiar Plan/i)],
        "Cambiar Plan button",
      );
    });

    await runStep("Detalles de la Cuenta", async () => {
      await waitForText(page, /Cuenta creada/i, "Cuenta creada");
      await waitForText(page, /Estado activo/i, "Estado activo");
      await waitForText(page, /Idioma seleccionado/i, "Idioma seleccionado");
    });

    await runStep("Tus Negocios", async () => {
      await waitForText(page, /Tus Negocios/i, "Tus Negocios section");
      await waitForAny(
        page,
        [() => page.getByRole("button", { name: /Agregar Negocio/i }), () => page.getByText(/Agregar Negocio/i)],
        "Agregar Negocio button",
      );
      await waitForText(page, /Tienes\s+2\s+de\s+3\s+negocios/i, "Business quota text");
    });

    await runStep("Términos y Condiciones", async () => {
      await validateLegalPage({
        appPage: page,
        linkText: "Términos y Condiciones",
        headingRegex: /Términos y Condiciones/i,
        reportField: "Términos y Condiciones",
        screenshotLabel: "05-terminos-y-condiciones",
      });
    });

    await runStep("Política de Privacidad", async () => {
      await validateLegalPage({
        appPage: page,
        linkText: "Política de Privacidad",
        headingRegex: /Política de Privacidad/i,
        reportField: "Política de Privacidad",
        screenshotLabel: "06-politica-de-privacidad",
      });
    });
  } finally {
    const report = {
      generatedAt: new Date().toISOString(),
      environment: {
        initialUrl: loginUrl || page.url(),
        currentUrl: page.url(),
      },
      results,
      legalUrls,
      screenshotPaths,
      errors: stepErrors,
    };

    const reportPath = path.join(ARTIFACTS_DIR, "final-report.json");
    await fs.writeFile(reportPath, JSON.stringify(report, null, 2), "utf8");

    console.log("\n=== FINAL REPORT ===");
    for (const field of REPORT_FIELDS) {
      console.log(`- ${field}: ${results[field]}`);
    }
    if (Object.keys(legalUrls).length > 0) {
      console.log("\nLegal URLs:");
      for (const [name, url] of Object.entries(legalUrls)) {
        console.log(`- ${name}: ${url}`);
      }
    }
    console.log(`\nReport path: ${reportPath}`);

    await context.close();
  }

  const hasFailures = Object.values(results).includes("FAIL");
  if (hasFailures) {
    process.exitCode = 1;
  }
}

main().catch((error) => {
  console.error("Fatal run error:", error);
  process.exit(1);
});
